pset/blinder.go

```diff
@@ -0,0 +1,314 @@
+) {
+	if err := pset.SanityCheck(); err != nil {
+		return nil, err
+	}
+
+	return &blinder{
+		pset:             pset,
+		blindingPrivkeys: blindingPrivkeys,
+		blindingPubkeys:  blindingPubkeys,
+	}, nil
+}
+
+//BlindOutputs method blinds pset's output
+func (b *blinder) BlindOutputs() error {
+	err := b.validate()
+	if err != nil {
+		return err
+	}
+
+	unblindOutputs, err := b.unblindInputs()
+	if err != nil {
+		return err
+	}
+
+	outputValues := make([]uint64, 0)
+	//TODO check if looping over UnsignedTx's Outputs is correct?
+	for _, output := range b.pset.UnsignedTx.Outputs {
+		var val [9]byte
+		copy(val[:], output.Value)
+		value, err := confidential.ElementsToSatoshiValue(val)
+		if err != nil {
+			return err
+		}
+		outputValues = append(outputValues, value)
+	}
+
+	inputAbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAbfs = append(inputAbfs, v.AssetBlindingFactor)
+	}
+
+	inputVbfs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputVbfs = append(inputVbfs, v.ValueBlindingFactor)
+	}
+
+	inputAgs := make([][]byte, 0)
+	for _, v := range unblindOutputs {
+		inputAgs = append(inputAgs, v.Asset)
+	}
+
+	inputValues := make([]uint64, 0)
+	for _, v := range unblindOutputs {
+		inputValues = append(inputValues, v.Value)
+	}
+
+	outputVbfs, outputAbfs, err := b.generateOutputBlindingFactors(
+		inputValues,
+		outputValues,
+		inputAbfs,
+		inputVbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = b.blindOutputs(
+		outputValues,
+		outputAbfs,
+		outputVbfs,
+		inputAgs,
+		inputAbfs,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *blinder) validate() error {
+	for _, input := range b.pset.Inputs {
+		if input.NonWitnessUtxo == nil && input.WitnessUtxo == nil {
+			return errors.New("all inputs must contain a non witness " +
+				"utxo or a witness utxo")
+		}
+	}
+
+	if len(b.blindingPrivkeys) != len(b.pset.Inputs) {
+		return errors.New("blinding private keys do not match the number" +
+			" of inputs")
+	}
+
+	if len(b.blindingPubkeys) != (len(b.pset.Outputs) - 1) {
+		return errors.New("blinding public keys do not match the number " +
+			"of outputs (fee excluded)")
+	}
+	return nil
+}
+
+func (b *blinder) unblindInputs() ([]confidential.UnblindOutputResult, error) {
+	ctx, _ := secp256k1.ContextCreate(secp256k1.ContextBoth)
+	defer secp256k1.ContextDestroy(ctx)
+	unblindOutputs := make([]confidential.UnblindOutputResult, 0)
+	for index, input := range b.pset.Inputs {
+		var prevout *transaction.TxOutput
+		if input.NonWitnessUtxo != nil {
+			//TODO: check if prevout should be created like below
+			prevout = input.NonWitnessUtxo.Outputs[index]
+		} else {
+			prevout = input.WitnessUtxo
+		}
+
+		if len(prevout.RangeProof) > 0 && len(prevout.SurjectionProof) > 0 {
+			val, err := secp256k1.CommitmentParse(ctx, prevout.Value)
+			if err != nil {
+				return nil, err
+			}
+			var commitmentValue secp256k1.Commitment
+			if val != nil {
+				commitmentValue = *val
+			}
```

Comment:
secp256k1.CommitmentParse returns *Commitment.
ValueCommit(UnblindInput) is of type Commitment, since I need to dereference I wanted to prevent null pointer panic in case secp256k1.CommitmentParse returns nil.
I guess its a small chance that this can happen but wanted to be on the safe side.
Do you want me to remove it?

Reply:
`secp256k1.CommitmentParse` should return a nil commitment only in case an error has occurred. Although I understand the concern, I think we should be good without this.

What's really annoying for me, is that we do not have a `CommitmentFromBytes` to avoid managing the creation and destruction of a context here. We'll fix this, if you could just  add a todo comment for this.

Reply:
Done